components/butterfly/src/server/push.rs

```diff
@@ -107,18 +109,17 @@ fn run_loop(server: &Server, timing: &Timing) -> ! {
                 let _ = guard.join()
                              .map_err(|e| error!("Push worker died: {:?}", e));
             }
-            if Instant::now() < next_gossip {
-                let wait_time = (next_gossip - Instant::now()).as_millis();
-                if wait_time > 0 {
-                    thread::sleep(Duration::from_millis(wait_time as u64));
-                }
+            // If we've still got any time left in the gossip period, sleep
+            // for that long.
+            if let Some(wait_time) = gossip_period.checked_sub(gossip_start_time.elapsed()) {
+                thread::sleep(wait_time);
             }
         }
-        if Instant::now() < long_wait {
-            let wait_time = (long_wait - Instant::now()).as_millis();
-            if wait_time > 0 {
-                thread::sleep(Duration::from_millis(wait_time as u64));
-            }
+
+        // If we've still got any time left in the gossip period, sleep
+        // for that long.
+        if let Some(wait_time) = gossip_period.checked_sub(fanout_loop_start_time.elapsed()) {
+            thread::sleep(wait_time);
         }
```

Comment:
This pattern is used in a few other places. Would it be worth factoring it out into a function with a descriptive name (`sleep_from_start_time`?)? It might make the code easier to parse at a glance and remove the need for comments.

Reply:
Yes! This is one of the additional cleanups I'm working on. I wanted to get the main fix out first, though, but the change you ask for will be coming soon (just maybe not in this PR) :smile: